### PR TITLE
Clear expired lightning payments

### PR DESCRIPTION
**Pull request by klochowicz**
See commit details for more details.

screenshots
image

payment details:
image

### CODE REVIEW

**Comment by klochowicz** [comment]
note: I'm not dealing with invoices created before this changeset (they will have no expiry timestamp) - I will create a separate PR for them, as I suspect more discussions arising from it.

File: rust/src/db.rs [hunk 1]
```diff
@@ -338,6 +343,7 @@ mod tests {
             Flow::Inbound,
             crate::lightning::HTLCStatus::Pending,
             MillisatAmount(Some(100000)),
+            Some(get_timestamp() + 1), // 2 secs expiry time
```
Comment (luckysori) [inline rust/src/db.rs]: ❓ Is it not 1 sec?

File: rust/src/lightning.rs [hunk 2]
```diff
@@ -1152,6 +1156,7 @@ pub async fn send_payment(invoice: &Invoice) -> Result<()> {
         Flow::Outbound,
         status,
         MillisatAmount(invoice.amount_milli_satoshis()),
+        Some(invoice.duration_since_epoch().as_secs() + invoice.expiry_time().as_secs()),
```
Comment (luckysori) [inline rust/src/lightning.rs:1160]: A bit annoying that the library doesn't give you the timestamp directly!

File: rust/src/lightning.rs [hunk 3]
```diff
@@ -1001,6 +1004,7 @@ async fn handle_ldk_events(
                     Flow::Inbound,
                     HTLCStatus::Succeeded,
                     MillisatAmount(Some(*amount_msat)),
+                    None,
```
Comment (luckysori) [inline rust/src/lightning.rs:1008]: 🙃 It's weird to have to deal with this here because the payment has already succeeded.

File: rust/src/db.rs [hunk 4]
```diff
@@ -323,6 +324,36 @@ pub async fn load_payments() -> Result<Vec<PaymentInfo>> {
...
+                    }
+                })
+                .is_some()
+        })
+        .cloned()
+        .collect::<Vec<PaymentInfo>>();
+
+    for payment in newly_expired {
+        update_payment(&payment.hash, HTLCStatus::Failed, None).await?;
+    }
```
Comment (luckysori) [inline rust/src/db.rs:353]: 🙃 Using `iter_mut().for_each()` might be a bit simpler.

File: rust/src/db.rs [hunk 5]
```diff
@@ -323,6 +324,36 @@ pub async fn load_payments() -> Result<Vec<PaymentInfo>> {
...
+                        None
+                    }
+                })
+                .is_some()
+        })
+        .cloned()
+        .collect::<Vec<PaymentInfo>>();
+
+    for payment in newly_expired {
+        update_payment(&payment.hash, HTLCStatus::Failed, None).await?;
```
Comment (luckysori) [inline rust/src/db.rs]: 🙃 It could be nice to differentiate between failed payments and expired payments in the UI.

File: rust/src/db.rs [hunk 6]
```diff
@@ -353,9 +384,11 @@ mod tests {
             .set_default()
     }
 
-    async fn ensure_init_fresh_db() -> Result<()> {
+    // Prepare DB for the test. Each test should have DB with unique name to
+    // avoid race conditions.
```
Comment (luckysori) [inline rust/src/db.rs:388]: 🙃 What about a random name? It's harder to make a mistake then and easier to use.

File: rust/src/lightning.rs [hunk 7]
```diff
@@ -200,6 +200,7 @@ pub enum HTLCStatus {
     Pending,
     Succeeded,
     Failed,
+    Expired,
```
Comment (luckysori) [inline rust/src/lightning.rs:203]: 👍 Nice!

**Review by luckysori — APPROVED** [review]
Nice job! I don't think any of my comments are critical.

File: rust/src/db.rs (hunk 1 above)
Comment (luckysori) [reply rust/src/db.rs]: This changes in a latter commit anyway.

File: rust/src/lightning.rs (hunk 3 above)
Comment (klochowicz) [reply rust/src/lightning.rs:1008]: there's definitely potential for better types and better API here.
I believe the awkwardness is that you might receive a payment without creating an invoice beforehand, hence you only learn about it when you get it.

File: rust/src/db.rs (hunk 4 above)
Comment (klochowicz) [reply rust/src/db.rs:353]: I don't mind re-working it, I believe it's exactly the same under-the-hood for this syntactic sugar :)

File: rust/src/db.rs (hunk 6 above)
Comment (klochowicz) [reply rust/src/db.rs:388]: if there were more than 2 tests, then yes :P didn't want to invest more time than absolutely needed at this point ;D

**Review by holzeis — APPROVED** [review]
LGTM 👍

File: rust/src/db.rs (hunk 5 above)
Comment (klochowicz) [reply rust/src/db.rs]: I have added capability for this in my changeset :-)- you can diferrentiate it the payment details (a new icon can be added with a one-liner in Dart after this gets in, I just didn't come up with any)

File: rust/src/db.rs (hunk 4 above)
Comment (klochowicz) [reply rust/src/db.rs:353]: I don't think I can use `for_each` here, as async closures are still unstable....

Comment (luckysori) [reply rust/src/db.rs]: Yep, I didn't notice the `await` below! Definitely better to go for a for loop here.

**Pull request merged**